content/docs/iac/using-pulumi/extending-pulumi/build-a-provider.md

```diff
@@ -0,0 +1,722 @@
+		return infer.UpdateResponse[FileState]{}, nil
+	}
+
+	_, err := os.Stat(req.Inputs.Path)
+	if req.State.Content != req.Inputs.Content || os.IsNotExist(err) {
+		f, err := os.Create(req.State.Path)
+		if err != nil {
+			return infer.UpdateResponse[FileState]{}, err
+		}
+		defer f.Close()
+		n, err := f.WriteString(req.Inputs.Content)
+		if err != nil {
+			return infer.UpdateResponse[FileState]{}, err
+		}
+		if n != len(req.Inputs.Content) {
+			return infer.UpdateResponse[FileState]{}, fmt.Errorf("only wrote %d/%d bytes", n, len(req.Inputs.Content))
+		}
+	}
+
+	return infer.UpdateResponse[FileState]{
+		Output: FileState{
+			Path:    req.Inputs.Path,
+			Force:   req.Inputs.Force,
+			Content: req.Inputs.Content,
+		},
+	}, nil
+}
+```
+
+##### The `Diff` operation
+
+The `Diff` operation compares a resource's recorded state (if any) to the current input values for the resource, to determine what kind of changes need to be made. The `diff` object is a map of property names to the kind of diff operation (if any) that a property needs to have made. The logic of this function is straightforward: compare `req.Inputs.<propertyName>` to `req.State.<propertyName>` and if they aren't equal, add to the diff that this property needs to be updated using `p.PropertyDiff{Kind: p.Update}`. Only include the properties that need to change in the `diff` map. Finally return a `DiffResponse` containing the `diff` map, and set a few other options to configure the update behavior.
+
+```go
+func (File) Diff(ctx context.Context, req infer.DiffRequest[FileArgs, FileState]) (infer.DiffResponse, error) {
+	diff := map[string]p.PropertyDiff{}
+	if req.Inputs.Content != req.State.Content {
+		diff["content"] = p.PropertyDiff{Kind: p.Update}
+	}
+	if req.Inputs.Force != req.State.Force {
+		diff["force"] = p.PropertyDiff{Kind: p.Update}
+	}
+	if req.Inputs.Path != req.State.Path {
+		diff["path"] = p.PropertyDiff{Kind: p.UpdateReplace}
+	} else {
+		_, err := os.Stat(req.Inputs.Path)
+		if os.IsNotExist(err) {
+			diff["path"] = p.PropertyDiff{Kind: p.Add}
+		}
+	}
+	return infer.DiffResponse{
+		DeleteBeforeReplace: true,
+		HasChanges:          len(diff) > 0,
+		DetailedDiff:        diff,
+	}, nil
+}
+```
+
+#### The `Read` operation
+
+The `Read` operation fetches the resource. The `ReadRequest` has a `ID` property that can be used, in this case, to determine the path to the file, and the base library functions can be used to read the file from disk, populating the `Content` field.
+
+```go
+func (File) Read(ctx context.Context, req infer.ReadRequest[FileArgs, FileState]) (infer.ReadResponse[FileArgs, FileState], error) {
+	path := req.ID
+	byteContent, err := os.ReadFile(path)
+	if err != nil {
+		return infer.ReadResponse[FileArgs, FileState]{}, err
+	}
+	content := string(byteContent)
+	return infer.ReadResponse[FileArgs, FileState]{
+		ID: path,
+		Inputs: FileArgs{
+			Path:    path,
+			Force:   req.Inputs.Force && req.State.Force,
```

Comment:
Wouldn't it be more correct for `Force` to be purely the live state? Why would it consider the inputs at this stage?
In general, for drift detection to work as expected, read should return the actual state, not influenced by the goal state.

```go
Force:   req.State.Force,
```

Reply:
nice catch!